### PR TITLE
feature: add blis and other BLAS implementation support

### PR DESCRIPTION
**Pull request by zenixls2**
Library: https://github.com/flame/blis  (master)

Compare to openblas (focal-updates,now 0.3.8+ds-1ubuntu0.20.04.1 amd64) in benchmark:

hardware setup:
- 12th Gen Intel(R) Core(TM) i7-12700  (12 core 20 threads)
- 16GiB DIMM Synchronous 4800 MHz * 2

openblas-openmp (OMP_NUM_THREADS=14)

```
Matrix Multiplication of (11008,4096,1) x (11008,128,1) - about  11.54 gFLOPS

Iteration;NThreads; SizeX; SizeY; SizeZ; Required_FLOPS; Elapsed_u_Seconds; gigaFLOPS
=====================================================================================
        0;       1; 11008;  4096;   128;    11542724608;             81141;    142.26
        1;       1; 11008;  4096;   128;    11542724608;             56359;    204.81
        2;       1; 11008;  4096;   128;    11542724608;             55481;    208.05
        3;       1; 11008;  4096;   128;    11542724608;             56389;    204.70
        4;       1; 11008;  4096;   128;    11542724608;             55802;    206.85
        5;       1; 11008;  4096;   128;    11542724608;             56172;    205.49
        6;       1; 11008;  4096;   128;    11542724608;             56069;    205.87
        7;       1; 11008;  4096;   128;    11542724608;             56091;    205.79
        8;       1; 11008;  4096;   128;    11542724608;             56036;    205.99
        9;       1; 11008;  4096;   128;    11542724608;             56178;    205.47

```

blis (openmp, BLIS_NUM_THREADS=14)

```
cgraph->n_threads=1
Matrix Multiplication of (11008,4096,1) x (11008,128,1) - about  11.54 gFLOPS

Iteration;NThreads; SizeX; SizeY; SizeZ; Required_FLOPS; Elapsed_u_Seconds; gigaFLOPS
=====================================================================================
        0;       1; 11008;  4096;   128;    11542724608;             67620;    170.70
        1;       1; 11008;  4096;   128;    11542724608;             37163;    310.60
        2;       1; 11008;  4096;   128;    11542724608;             34049;    339.00
        3;       1; 11008;  4096;   128;    11542724608;             37200;    310.29
        4;       1; 11008;  4096;   128;    11542724608;             37662;    306.48
        5;       1; 11008;  4096;   128;    11542724608;             37220;    310.12
        6;       1; 11008;  4096;   128;    11542724608;             37129;    310.88
        7;       1; 11008;  4096;   128;    11542724608;             37251;    309.86
        8;       1; 11008;  4096;   128;    11542724608;             37341;    309.12
        9;       1; 11008;  4096;   128;    11542724608;             37738;    305.86

Average                                                                        298.29
=====================================================================================
```

mkl blas (Intel10_64lp)  (OMP_NUM_THREADS=14)
```

Iteration;NThreads; SizeX; SizeY; SizeZ; Required_FLOPS; Elapsed_u_Seconds; gigaFLOPS
=====================================================================================
        0;       1; 11008;  4096;   128;    11542724608;             64675;    178.47
        1;       1; 11008;  4096;   128;    11542724608;             38275;    301.57
        2;       1; 11008;  4096;   128;    11542724608;             38270;    301.61
        3;       1; 11008;  4096;   128;    11542724608;             38418;    300.45
        4;       1; 11008;  4096;   128;    11542724608;             38323;    301.20
        5;       1; 11008;  4096;   128;    11542724608;             38446;    300.23
        6;       1; 11008;  4096;   128;    11542724608;             38306;    301.33
        7;       1; 11008;  4096;   128;    11542724608;             38396;    300.62
        8;       1; 11008;  4096;   128;    11542724608;             38417;    300.46
        9;       1; 11008;  4096;   128;    11542724608;             38363;    300.88

Average                                                                        288.68
=====================================================================================

```

Also looking forward if someone else could make use of it's typed api to do further optimization.

### CODE REVIEW

**Comment by zenixls2** [comment]
GGML_USE_BLIS is not actually been used.
BLIS specific optimisation might be applied based on this.

**Comment by zenixls2** [comment]
I've noticed that https://github.com/ggerganov/whisper.cpp/pull/927 has some implementation to let other options possible
However, the cmake_minimum_required version might have to be upgraded, as
- `BLA_PKGCONFIG_BLAS`  New in version 3.25.
- `BLA_SIZEOF_INTEGER`    New in version 3.22.

Will change the title and modify the changes to align with whisper's pr.
cc @ggerganov

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.12) # Don't bump this version for no reason
+cmake_minimum_required(VERSION 3.25) # Don't bump this version for no reason
```
Comment (ggerganov) [inline CMakeLists.txt:1]: Let's drop the CMake changes - v3.25 is not easily accessible yet

**Review by ggerganov — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (zenixls2) [reply CMakeLists.txt:1]: Should we also drop BLA_SIZEOF_INTEGER?
So that current version has no need to change

Comment (zenixls2) [reply CMakeLists.txt:1]: Ah, let me add some if condition on the cmake version

File: CMakeLists.txt [hunk 2]
```diff
@@ -145,36 +146,28 @@ if (APPLE AND LLAMA_ACCELERATE)
...
 
-if (LLAMA_OPENBLAS)
+if (LLAMA_BLAS)
     if (LLAMA_STATIC)
         set(BLA_STATIC ON)
     endif()
-
-    set(BLA_VENDOR OpenBLAS)
+    if ($(CMAKE_VERSION) VERSION_GREATER_EQUAL 3.22)
+        set(BLA_SIZEOF_INTEGRER 8)
```
Comment (ggerganov) [inline CMakeLists.txt:154]:
```suggestion
        set(BLA_SIZEOF_INTEGER 8)
```

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
Sorry, I forgot to run the CI - we need to fix the builds before merging
Please reopen a new PR

**Pull request merged**